test/hotspot/gtest/gc/g1/test_g1BatchedGangTask.cpp

```diff
@@ -0,0 +1,157 @@
+  static void run_task(AbstractGangTask* task) {
+    work_gang()->run_task(task);
+  }
+};
+
+WorkGang* G1BatchedGangTaskWorkers::_work_gang = NULL;
+
+class G1TestSubTask : public G1AbstractSubTask {
+  mutable uint _phase;
+  volatile uint _num_do_work; // Amount of do_work() has been called.
+
+  void check_and_inc_phase(uint expected) const {
+    ASSERT_EQ(_phase, expected);
+    _phase++;
+  }
+
+  bool volatile* _do_work_called_by;
+
+protected:
+  uint _max_workers;
+
+  void do_work_called(uint worker_id) {
+    Atomic::inc(&_num_do_work);
+    bool orig_value = Atomic::cmpxchg(&_do_work_called_by[worker_id], false, true);
+    ASSERT_EQ(orig_value, false);
+  }
+
+  void verify_do_work_called_by(uint num_workers) {
+    ASSERT_EQ(Atomic::load(&_num_do_work), num_workers);
+    // Do not need to check the _do_work_called_by array. The count is already verified
+    // by above statement, and we already check that a given flag is only set once.
+  }
+
+public:
+  // Actual use of GCParPhasesSentinel will cause an assertion failure when trying
+  // to add timing information - this should be disabled here.
+  G1TestSubTask() : G1AbstractSubTask(G1GCPhaseTimes::GCParPhasesSentinel),
+    _phase(0),
+    _num_do_work(0),
+    _do_work_called_by(NULL),
+    _max_workers(0) {
+    check_and_inc_phase(0);
+  }
+
+  ~G1TestSubTask() {
+    check_and_inc_phase(3);
+    FREE_C_HEAP_ARRAY(bool, _do_work_called_by);
+  }
+
+  double num_busy_workers() const override {
+    check_and_inc_phase(1);
+    return 1.0;
+  }
+
+  // Called by G1BatchedGangTask to provide information about the the maximum
+  // number of workers for all subtasks after it has been determined.
+  void set_max_workers(uint max_workers) override {
+    assert(max_workers >= 1, "must be");
+    check_and_inc_phase(2);
+
+    _do_work_called_by = NEW_C_HEAP_ARRAY(bool, max_workers, mtInternal);
+    for (uint i = 0; i < max_workers; i++) {
+      _do_work_called_by[i] = false;
+    }
+    _max_workers = max_workers;
+  }
+
+  void do_work(uint worker_id) override {
+    do_work_called(worker_id);
+  }
+};
+
+class G1SerialTestSubTask : public G1TestSubTask {
+public:
+  G1SerialTestSubTask() : G1TestSubTask() { }
+  ~G1SerialTestSubTask() {
+    verify_do_work_called_by(1);
+  }
+
+  double num_busy_workers() const override {
+    G1TestSubTask::num_busy_workers();
+    return 1.0;
+  }
+};
+
+class G1ParallelTestSubTask : public G1TestSubTask {
+public:
+  G1ParallelTestSubTask() : G1TestSubTask() { }
+  ~G1ParallelTestSubTask() {
+    verify_do_work_called_by(_max_workers);
+  }
+
+  double num_busy_workers() const override {
+    G1TestSubTask::num_busy_workers();
+    return 2.0;
+  }
+};
+
+class G1TestBatchedGangTask : public G1BatchedGangTask {
+public:
+  G1TestBatchedGangTask() : G1BatchedGangTask("Batched Gang Test Task", NULL) {
+    _serial_tasks.push(new G1SerialTestSubTask());
+    _parallel_tasks.push(new G1ParallelTestSubTask());
```

Comment:
I suggest we add methods for adding serial and parallel tasks to the `G1BatchedGangTask`interface.